### PR TITLE
Implementation

### PR DESCRIPTION
**Pull request by Gabbendorf**

### CODE REVIEW

File: lib/hangman_rules.rb [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  @available_letters = ("a".."z").to_a
+  @right_guesses = []
+  end
+
+  def random_word()
+    words.sample()
+  end
+
+  def keep_track_of_guesses(secret_word, guessed_letter)
+    add_if_right_guess(secret_word, guessed_letter)
```
Comment (felipesere) [inline lib/hangman_rules.rb]: It's a bit odd to see a "_if_" in a method name.
What is the _intent_ of function? What do you need to know from the name here?
How is "add_if_right_guess" a step towards _keeping track of guesses_?
Also, the remove(..) function has no _if_ in the name, but does `.delete_if { ...}` 😄

File: lib/hangman_rules.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    remove(guessed_letter)
+  end
+
+  def get_letters()
```
Comment (felipesere) [inline lib/hangman_rules.rb]: In Ruby we rarely name things "get" and you already have a way to access the `available_letters` with the `attr_reader`

File: lib/word_formatter.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+  attr_reader :rules
+
+  def initialize(rules)
+    @rules = rules
+  end
+
+  def format(secret_word)
+    formatted_word = secret_word.split("").map do |letter|
+      letter = replaced_letter(letter)
```
Comment (felipesere) [inline lib/word_formatter.rb]: I am pretty sure you don't need this assignment here.
`.map(..)` will create a new array with whatever comes out of the `do |..| .. end` block

File: lib/word_formatter.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    end
+    formatted_word.join("")
+  end
+
+  private
+
+  def replaced_letter(letter)
+    if not_guessed?(letter)
+      letter = "_"
```
Comment (felipesere) [inline lib/word_formatter.rb]: Same here. You can write something as simple as:
```
if not_guessed?(letter)
  "_"
else
  letter
end
```

File: lib/hangman_rules.rb (hunk 1 above)
Comment (Gabbendorf) [reply lib/hangman_rules.rb]: Yeah, that needs to be changed

File: lib/words.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
+class Words
+
+  def get_words()
+    words.map {|key, value| value}
```
Comment (felipesere) [inline lib/words.rb]: Why do you need a key/value map with `:word_1` etc if you only use the value side?
An array would serve the same purpose.
And the same "get_" prefix we don' t traditionally use 😺

File: lib/hangman_rules.rb (hunk 2 above)
Comment (Gabbendorf) [reply lib/hangman_rules.rb]: Gotcha

File: lib/word_formatter.rb (hunk 3 above)
Comment (Gabbendorf) [reply lib/word_formatter.rb]: Even though I'm calling `join()` on it further below?

File: spec/hangman_rules_spec.rb [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+  it "returns a random word" do
+    hangman_rules = new_hangman_rules(WordsStub.new)
+
+    word = hangman_rules.random_word()
+
+    expect(word).to eq("hello")
+  end
+
+  it "stores letters correctly guessed" do
```
Comment (felipesere) [inline spec/hangman_rules_spec.rb]: "storing" is such a boring word.
how about "it remembers correctly guessed letters"?

File: lib/word_formatter.rb (hunk 4 above)
Comment (Gabbendorf) [reply lib/word_formatter.rb]: True

File: lib/words.rb (hunk 5 above)
Comment (Gabbendorf) [reply lib/words.rb]: I don't know, I was trying to use a module here but without success...

File: spec/hangman_rules_spec.rb [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    hangman_rules = new_hangman_rules(words)
+
+    hangman_rules.keep_track_of_guesses("hello", "h")
```
Comment (felipesere) [inline spec/hangman_rules_spec.rb]: Hmmm... `"hello"` here feels odd. I was kinda expecting that secret word to already be kept somewhere in the bowels of `hangman_rules`?

something like 
```
hangman_rules = new HangManRules(wordsStub)
hangman_rules.guess("h")
hangman_rules.guess("e)

expect(hangman_rules.correct_guesses).to eq(["h", "e"])
expect(hangman_rules.remaining_guesses).to eq(1_000_000)
```

File: spec/hangman_rules_spec.rb [hunk 8]
```diff
@@ -0,0 +1,61 @@
+require 'spec_helper'
+require_relative '../lib/hangman_rules'
+
+RSpec.describe HangmanRules do
```
Comment (felipesere) [inline spec/hangman_rules_spec.rb:4]: I think this is the key set of tests/facts about the game.
I would suggest you revisit these and make sure these read like how you'd explain the rules to me while
having a coffee/tea without a laptop.

File: spec/word_formatter_spec.rb [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+  let(:rules) {HangmanRules.new(words)}
+  let(:formatter) {WordFormatter.new(rules)}
+
+  it "returns word showing all lines for no letters guessed" do
+    formatted_word = formatter.format("hello")
+
+    expect(formatted_word).to eq("_____")
+  end
+
+  it "returns word showing lines and right guessed letters" do
```
Comment (felipesere) [inline spec/word_formatter_spec.rb]: here I'd write something like "correctly guessed letters are revealed"

File: lib/word_formatter.rb (hunk 3 above)
Comment (felipesere) [reply lib/word_formatter.rb]: Yeah, you can chain the join() into it.
like this:
```
secret_words.split("").map { |letter| hide_or_reveal(letter) }.join("")
```

File: lib/words.rb (hunk 5 above)
Comment (felipesere) [reply lib/words.rb]: Its ok. It's always a pleasure to simplify code 🛩

File: spec/hangman_rules_spec.rb (hunk 7 above)
Comment (Gabbendorf) [reply spec/hangman_rules_spec.rb]: You're right!! My app is missing a `guess` method!!!

**Pull request merged**